[SPMD] Move some functions to IterGraphModule.setup()

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99076
* #99075
* #99069

Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

fegin committed:
[SPMD] Move some functions to IterGraphModule.setup()

Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Move some functions to IterGraphModule.setup()"


Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Move some functions to IterGraphModule.setup()"


Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Move some functions to IterGraphModule.setup()"


Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

[ghstack-poisoned]

lessw2020 commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -33,12 +32,12 @@ def __call__(self, gm: fx.GraphModule) -> Callable:
             remove_copy_from_optimizer(iter_gm)
-        iter_gm.freeze_cross_iter_movement()
+        # Must be called after we are not going to move the graphs
         iter_gm.setup(self.num_iters)
```
future idea not related to this PR directly - seems like 'finalize' would be clearer name than setup here (setup implies first step, vs this is last step before locking down graphs and also this means having multiple setup references - could be confused with the 'setup' graph as an entity vs this as a function).

fegin replied:
Thanks for the suggestion. Right now we are unable to change the naming as it is already used by some internal code. Will change this accordingly afterward!

lessw2020 reviewed (approved):
looks good.

fegin committed:
Update on "[SPMD] Move some functions to IterGraphModule.setup()"


Since users will have to call these steps before calling `setup()`, moving these steps to `setup()` can reduce the API usage complexity.

Differential Revision: [D44973726](https://our.internmc.facebook.com/intern/diff/D44973726/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
